File: dev/sources/distro/Src-Utilities/tileWin/tileWin.cpp
/*********************************************************************
tileWin
Copyright (C) 2004-2013 Roberto Rossi 
Web : http://www.redchar.net

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*********************************************************************/

// Questa applicazione, dato il parziale del titolo di una
// finestra, prende tutte le finestre con quelsto titolo e
// le affianca in orizzontale o verticale

#include <windows.h>
#include <string>
#include "tileWin.h"

#define SIZE_LIST 100
#define SIZE 300
#define FIND_APP_NAME " SciTE"

HWND local_windows_list[SIZE_LIST]; //contiene l'elenco delle finestre trovate
int  local_current_index;


void print_info(void)
{
  std::string msg;

  msg.append("*********************************************************************");
  msg.append("\ntileWin");
  msg.append("\nCopyright (C) 2004-2013 Roberto Rossi ");
  msg.append("\nWeb : http://www.redchar.net\n");

  msg.append("\nThis library is free software; you can redistribute it and/or");
  msg.append("\nmodify it under the terms of the GNU Lesser General Public");
  msg.append("\nLicense as published by the Free Software Foundation; either");
  msg.append("\nversion 2.1 of the License, or (at your option) any later version.\n");

  msg.append("\nThis library is distributed in the hope that it will be useful,");
  msg.append("\nbut WITHOUT ANY WARRANTY; without even the implied warranty of");
  msg.append("\nMERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU");
  msg.append("\nLesser General Public License for more details.\n");

  msg.append("\nYou should have received a copy of the GNU Lesser General Public");
  msg.append("\nLicense along with this library; if not, write to the Free Software");
  msg.append("\nFoundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA");
  msg.append("\n*********************************************************************");

  MessageBox(NULL, msg.c_str(), "Info", MB_OK);
}


//ritorna il numero di finestre da affiancare, presenti in local_windows_list
int count_windows()
{
  int i;
  int c=0;

  for(i=0;i<SIZE_LIST;i++)
  {
    if (local_windows_list[i]==0){
      break;
    }
    c++;
  }

  return c;
}

//esegue l'affiancamento delle finestre
void execute_tile(int type_tile)
{
  int i;
  int nw;
  int xvid, yvid; //dimensioni video
  int x,y,w,h; //dimensioni
  HWND hwndWin;
  RECT r;

  nw = count_windows(); //finestre da affiancare
  if (nw > 1) //se esistono più finestre
  {
    SystemParametersInfo(SPI_GETWORKAREA, 0, &r, 0);

    // width
    xvid = r.right;
    // height
    yvid = r.bottom;

    if (type_tile == 0) //allineamento orizzontale
    {
      w = xvid / nw;
      h = yvid;
    } else {
      w = xvid;
      h = yvid / nw;
    }
    
    x = r.left;
    y = r.top;

    for( i=0 ; i < nw ; i++) //ciclo di affiancamento
    {
      hwndWin=local_windows_list[i];
      if (type_tile == 0) //allineamento orizzontale
      {
        x = w * i;
      } else { //allineamento verticale
        y = h * i;
      }
      if (IsZoomed(hwndWin) != 0)
        ShowWindow(hwndWin,SW_RESTORE);
      MoveWindow(hwndWin,x,y,w,h,TRUE);
      SetWindowPos(hwndWin,HWND_TOPMOST,x,y,w,h,SWP_NOMOVE | SWP_NOSIZE | SWP_NOREPOSITION);
      SetWindowPos(hwndWin,HWND_NOTOPMOST,x,y,w,h,SWP_NOMOVE | SWP_NOSIZE | SWP_NOREPOSITION);
      ShowWindow(hwndWin,SW_SHOW);
    }
  }//endif
}

//reset array generale per la lista finestre
void reset_windows_list()
{
  int i;

  for(i=0;i<SIZE_LIST;i++)
  {
    local_windows_list[i]=0;
  }
}

BOOL CALLBACK EnumWndProc(HWND hWnd,LPARAM lParam){
 char pTitle[SIZE];
 char buffer[1000];

 GetWindowText(hWnd,pTitle,SIZE);
 GetClassName(hWnd, buffer, sizeof(buffer));
 if (strstr(pTitle, FIND_APP_NAME))
 {
	if (strstr(buffer, "SciTEWindow"))
	{
   //h = GetParent(hWnd);
   //if (h == 0) {
    local_windows_list[local_current_index]=hWnd;
    local_current_index++;
	}
   //}
 }
 
 return TRUE;
}

int APIENTRY WinMain(HINSTANCE hInstance,
                     HINSTANCE hPrevInstance,
                     LPTSTR    lpCmdLine,
                     int       nCmdShow)
{
  int type_tile = -1;
  local_current_index = 0;

  if (strcmp(lpCmdLine,"-v") == 0)
    type_tile = 1;

  if (strcmp(lpCmdLine,"-o") == 0)
    type_tile = 0;

  if (type_tile > -1)
  {
    reset_windows_list();
    EnumWindows(EnumWndProc,0);
    execute_tile(type_tile);
  } else
    print_info();

	return (int) 1;
}

File: sources/distro/luascr/tilewin.lua
--[[
Version : 1.2.0
Web     : http://www.redchar.net

Questa procedura Consente la gestione dell'affiancamento delle finestre,
in modo da poter visualizzare contemporaneamente più file.

Copyright (C) 2004-2009 Roberto Rossi 
*******************************************************************************
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
*******************************************************************************
]]

do
  
  require("luascr/rluawfx")

  --Questa funzione controlla l'avvenuta apertura di una nuova istanza di SciTE
  --Se resetFlag = true, si limita a resettare il file che segnala l'apertura
  --di SciTE
  local function attendiAperturaSciTE(resetFlag)
    local nomef = ""
    local idf
    local res = false
    local i = 0
    
    nomef = os.getenv("TMP")
    nomef = nomef.."\\scitePrj.tmp"
    
    if (resetFlag) then      
      os.remove(nomef)
      --print("\nCancellato file temporaneo")
    else --attende apertura programma
      while (i < 500) do 
        rwfx_Sleep(200)
        if (rfx_fileExist(nomef)) then
          res = true
          --print("\nTrovato file temporaneo")
          break
        end
        i = i + 1
      end
    end
    
    return res
  end
  
  --controlla la presenza di almeno due file aperti,
  --ritorna false se è aperto un solo file
  local function controlloFileAperti()
    local res = true
    local buffers
    
    buffers = PUBLIC_get_bufferList()
    if (table.getn(buffers) < 2) then
      rwfx_MsgBox("Impossibile eseguire il comando. Attualmente è presente un solo file, risulta quindi inutile riaprirlo in una nuova finestra.","Attenzione")
      res = false
    end
    
    return res
  end
  
  --controlla la corretta impostazione della variabile relativa alle finestre
  --multiple e, se è il caso, avvisa l'utente delle modifiche necessarie.
  --Ritorna true se è tutto ok
  local function controlloStato()
    local res = true
    local edit
    local pos
    local linean
    local idm_bookmark_toggle = 222
    
    if (props["check.if.already.open"] == "1") then    
      --[[edit = rwfx_MsgBox("Impossibile continuare! \n\r\n\r"..
                  "Questa funzione necessità che SciTE lavori in modalità multi-finestra. \n\r"..
                  "Attualmente il programma funziona in modalità a finestra singola, è quindi necessario "..
                  "modificare l'opzione 'check.if.already.open'. \n\r\n\r"..
                  "Per fare ciò aprire il file 'SciTEGlobal.properties' e impostare "..
                  "la variabile 'check.if.already.open' a 0 (attualmente è impostata a 1).\n\r\n\r"..  
                  "Si desidera aprire automaticamente il file per la modifica?",
                  "Attenzione.",MB_ICONSTOP + MB_YESNO)
        ]]
      edit = rwfx_MsgBox(_t(138),_t(9),MB_ICONSTOP + MB_YESNO)
      if (edit == IDYES) then --procedere all'editazione
        --scite.Open(props["SciteDefaultHome"].."/SciTEGlobal.properties")
        scite.Open(props["SciteUserHome"].."/SciTEUser.properties")
        pos = editor:findtext("check.if.already.open",SCFIND_WHOLEWORD,0)
        if (pos) then --apertura file e selezione voce
          editor:GotoPos(pos)
          scite.MenuCommand(idm_bookmark_toggle) --evidenzia valore
        else
          --[[ 1 = permetti una sola finestra di SciTE
               0 = consenti la presenza di pià finestre di SciTE
               check.if.already.open=0
            ]]
          editor:AddText(_t(188))
        end
      end
      res = false
    end
    
    
    return res
  end

  --funzione principale
  local function main()
    local scelta
    local listaFxStr
    local nomef
    local comando
    local comandoSciTE
    local ultimo
    local i
    
    if (controlloStato()) then
      comando = props["SciteDefaultHome"].."/tilewin.exe"
      comandoSciTE = props["SciteDefaultHome"].."/SciTE.exe"
      
      --[[listaFxStr = 
                   "Affianca orizzontalmente le finestre di SciTE|"..   --0
                   "Affianca verticalmente le finestre di SciTE"        --1
                   "Apri nuova finestra|"..          --2
                   "Sposta file corrente in nuova finestra|"..            --3
                   "Apri file corrente in nuova finestra|"..            --4
        ]]
      listaFxStr = _t(137)
      
      scelta = rwfx_ShowList(listaFxStr,"Affiancamento SciTE...")
      if scelta then
        nomef = props["FilePath"] --file corrente
        if (scelta==3) then --Sposta file corrente in nuova finestra
          if (controlloFileAperti()) then
            attendiAperturaSciTE(true)
            rwfx_ShellExecute(comandoSciTE,"\""..nomef.."\"")
            attendiAperturaSciTE(false)
            rwfx_ExecuteCmd("close:")
          end
        elseif (scelta==2) then --Apri nuova finestra
          if (controlloFileAperti()) then
            rwfx_ShellExecute(comandoSciTE,"")
          end
        elseif (scelta==4) then --Apri file corrente in nuova finestra
          if (controlloFileAperti()) then
            rwfx_ShellExecute(comandoSciTE,"\""..nomef.."\"")
          end
        elseif (scelta==0) then --affiancamento orizzontale
          rwfx_ShellExecute(comando,"-o") --affiancamento orizzontale
        elseif (scelta==1) then --affiancamento verticale
          rwfx_ShellExecute(comando,"-v") --affiancamento verticale
        --elseif (scelta==4) then --apri tutti i file correnti
        --  --TODO : implementare affiancamento verticale
        --    for i,nomef in ipairs(buffers) do
        --      attendiAperturaSciTE(true)
        --      rwfx_ShellExecute(comandoSciTE,nomef)
        --      attendiAperturaSciTE(false)
        --    end
        --   rwfx_ShellExecute(comando,"") --affiancamento orizzontale
        --    rwfx_ExecuteCmd("quit:")
        end
      end
    end -- end controllo stato
  end --end function
  
  main()
end
